src/vmm/src/config.rs

```diff
@@ -1,13 +1,84 @@
 // Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
 
+use std::collections::HashMap;
 use std::convert::TryFrom;
 use std::fmt;
+use std::num::{NonZeroU8, ParseIntError};
 use std::path::PathBuf;
 use std::result;
 
 use super::{DEFAULT_HIGH_RAM_START, DEFAULT_KERNEL_CMDLINE};
 
+trait ArgParsing {
+    fn parse(raw: String) -> Result<Self, ParseIntError>
+    where
+        Self: Sized;
+}
+
+macro_rules! argparsing{
+    ($($t:ty)*) => ($(impl ArgParsing for $t {
+        fn parse(raw: String) -> Result<Self, ParseIntError> {
+            raw.parse()
+        }
+    })*)
+}
+
+argparsing! { u32 u64 NonZeroU8 }
+
+impl ArgParsing for PathBuf {
+    fn parse(raw: String) -> Result<Self, ParseIntError> {
```

Comment:
May be define a `ParseError` `Enum` and have different variants?

Reply:
Ignore above comment, I see - that's a `trait` not a `struct`, but still the error type be a different Error. May be an associated Error type that is an `enum` and that uses different variants?

Reply:
I already have added a `ParsingError` enum with different variants. 

[Here](https://github.com/rust-vmm/vmm-reference/pull/56/files#diff-919306a64a3d1ead59ffc1193399bb62123b222fc3a748b5d9dc3a34d6bdd6bfR60) I'm taking the possible error from the parse function, and wrap it in a ParsingError variant which also holds the parameter and the error. To do what you suggested, I would need to create yet another enum variant only for the parse function. I decide to use the `num::ParseIntError` as error type instead, because numeric conversion are the only parsing which can fail (parsing String -> String and String -> PathBuf is infallible).

If you think it's better, I can use the ParsingError variant, and pass the param_name as input to the parse function. In this way, each `parse` implementation can use it for building an helpful error. WDYT?

Reply:
Yes, I still can't think of what is a 'right' way, but something that returned a `String` as a result, returning `ParseIntError` didn't seem too right for me. So this `ParsingError` should be fine for now.